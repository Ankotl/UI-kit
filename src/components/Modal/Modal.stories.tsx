import React from 'react'

import { useToggle } from '../../hooks/useToggle/useToggle'
import { getStoryMeta } from '../../utils/storybook/getStoryMeta'
import { Button } from '../Button'
import { EnumTypographyVariants, Typography } from '../Typography/Typography'

import { Modal as ModalCmp, IModalProps } from './Modal'

const storiesMeta = getStoryMeta({
  component: ModalCmp,
})
export default storiesMeta

/* Interactive */
export const Modal = (
  props: Omit<IModalProps, 'isOpen' | 'onClose' | 'classes'>
) => {
  const modalToggler = useToggle()

  const containerStyle = {
    display: 'flex',
    justifyContent: 'center',
    alignItem: 'center',
    height: '100vh',
  }

  return (
    <div style={containerStyle}>
      <Button onClick={modalToggler.toggle}>Toggle Modal</Button>

      <ModalCmp
        {...props}
        isOpen={modalToggler.value}
        onClose={modalToggler.unset}
      >
        <Typography
          variant={EnumTypographyVariants.H4}
          text="Lorem ipsum dolor sit amet consectetur adipisicing elit. Dignissimos quod delectus praesentium ea! Quos quisquam magni, repellat odio exercitationem ex?"
        />
      </ModalCmp>
    </div>
  )
}

Modal.argTypes = {
  isOpen: {
    table: {
      disable: true,
    },
  },
  onClose: {
    table: {
      disable: true,
    },
  },
  classes: {
    table: {
      disable: true,
    },
  },
}
