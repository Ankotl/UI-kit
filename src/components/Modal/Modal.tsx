import React, { useCallback } from 'react'

import { ReactFocusLockProps } from 'react-focus-lock/interfaces'

import { useDisableOverflow } from 'hooks/useDisableOverflow/useDisableOverflow'
import { useEventListener } from 'hooks/useEventListener/useEventListener'

import { ModalBlackout } from './_components/ModalBlackout/ModalBlackout'
import { ModalBody } from './_components/ModalBody/ModalBody'

export interface IModalClasses {
  bodyClassName?: string
  contentClassName?: string
  titleClassName?: string
  closeIconClassName?: string
  blackoutClassName?: string
}

export interface IModalProps {
  isOpen: boolean
  children?: React.ReactNode
  title?: string
  enableCloseIcon?: boolean
  isCloseDisabled?: boolean

  classes?: IModalClasses
  modalBodyPortalParent?: HTMLElement | null | undefined
  blackoutPortalParent?: HTMLElement | null | undefined
  isMobileFullScreen?: boolean
  focusLockProps?: ReactFocusLockProps

  bodyRef?: React.MutableRefObject<HTMLDivElement | null>

  onClose?: () => void
}

export const Modal: React.FC<IModalProps> = ({
  isOpen,
  title,
  enableCloseIcon,
  isCloseDisabled,
  classes,
  onClose,
  children,
  bodyRef,
  focusLockProps,
  modalBodyPortalParent,
  blackoutPortalParent,
  isMobileFullScreen,
}) => {
  const handleClose = useCallback(() => {
    if (onClose && !isCloseDisabled) {
      onClose()
    }
  }, [isCloseDisabled, onClose])

  const handleKeyUp = useCallback(
    (e: KeyboardEvent) => {
      if (isOpen && e.key === 'Escape') {
        e.preventDefault()
        e.stopPropagation()
        handleClose()
      }
    },
    [isOpen, handleClose]
  )

  useEventListener('keyup', handleKeyUp)
  useDisableOverflow(
    isOpen,
    undefined,
    blackoutPortalParent || modalBodyPortalParent || document.body
  )

  return (
    <>
      <ModalBody
        classes={classes}
        enableCloseIcon={enableCloseIcon}
        isOpen={isOpen}
        onClose={handleClose}
        title={title}
        isMobileFullScreen={isMobileFullScreen}
        modalBodyPortalParent={modalBodyPortalParent}
        bodyRef={bodyRef}
        focusLockProps={focusLockProps}
      >
        {children}
      </ModalBody>

      <ModalBlackout
        isOpen={isOpen}
        blackoutPortalParent={blackoutPortalParent}
        modalBodyPortalParent={modalBodyPortalParent}
        isCloseDisabled={isCloseDisabled}
        classes={classes}
        onClose={handleClose}
      />
    </>
  )
}
