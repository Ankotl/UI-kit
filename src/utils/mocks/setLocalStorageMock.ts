import { EnumLocalStorageKeys } from '../../enums/localStorageEnum'

export const setLocalStorageMock = () => {
  class LocalStorageMock {
    store: Record<string, string> = {}

    clear() {
      this.store = {}
    }

    setItem(key: EnumLocalStorageKeys, value: string) {
      this.store[key] = value
    }

    getItem(key: EnumLocalStorageKeys) {
      return this.store[key] || null
    }

    removeItem(key: EnumLocalStorageKeys) {
      delete this.store[key]
    }
  }

  Object.defineProperty(window, 'localStorage', {
    value: new LocalStorageMock(),
  })
}
