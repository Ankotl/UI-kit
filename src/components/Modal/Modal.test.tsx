import React from 'react'

import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import '../../utils/mocks/setWindowMatchMediaMock'

import { Modal } from './Modal'

describe('Modal component test', () => {
  test('Mounts correctly', () => {
    const { baseElement } = render(
      <Modal isOpen={true} title={'title'}>
        test
      </Modal>
    )

    expect(baseElement).toMatchSnapshot()
  })

  test('Unmounts correctly', () => {
    const { baseElement } = render(<Modal isOpen={false}>test</Modal>)
    expect(baseElement).toMatchSnapshot()
  })

  test('Closes correctly', () => {
    const onClose = jest.fn()
    const { getByTestId } = render(<Modal onClose={onClose} isOpen />)
    fireEvent.mouseDown(getByTestId('modal-blackout'))
    fireEvent.click(getByTestId('modal-body'))
    fireEvent.keyUp(window, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    })
    expect(onClose).toBeCalledTimes(2)
  })

  test('Disables correctly', () => {
    const onClose = jest.fn()
    const { getByTestId } = render(
      <Modal onClose={onClose} isOpen isCloseDisabled />
    )

    fireEvent.mouseDown(getByTestId('modal-blackout'))
    fireEvent.click(getByTestId('modal-body'))
    fireEvent.keyUp(window, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    })
    expect(onClose).toBeCalledTimes(0)
  })
})
