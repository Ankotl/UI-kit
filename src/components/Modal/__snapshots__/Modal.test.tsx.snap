// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component test Mounts correctly 1`] = `
<body
  style="overflow-y: hidden;"
>
  <div />
  <div
    class="ModalBody"
  >
    <div
      class="ModalBody__content"
      data-testid="modal-body"
      style="opacity: 0.25; transform: translateY(-100%);"
    >
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        class="ModalBody__focusLock"
        data-focus-lock-disabled="false"
      >
        <div
          class="Typography Typography_h2 ModalBody__title"
          data-testid="modal-title"
        >
          title
        </div>
        test
      </div>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="Blackout Blackout_pointer"
    data-testid="modal-blackout"
    style="opacity: 0;"
  />
</body>
`;

exports[`Modal component test Unmounts correctly 1`] = `
<body
  style=""
>
  <div />
</body>
`;
